compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -935,38 +937,55 @@ impl<'p, 'tcx> Usefulness<'p, 'tcx> {
         matrix: &Matrix<'p, 'tcx>, // used to compute missing ctors
         ctor: &Constructor<'tcx>,
         ctor_wild_subpatterns: &Fields<'p, 'tcx>,
+        is_non_exhaustive: bool,
     ) -> Self {
+        let add_new_patterns = |new_patterns: Vec<Pat<'tcx>>, witnesses: Vec<Witness<'tcx>>| {
+            witnesses
+                .into_iter()
+                .flat_map(|witness| {
+                    new_patterns.iter().map(move |pat| {
+                        let mut witness = witness.clone();
+                        witness.0.push(pat.clone());
+                        witness
+                    })
+                })
+                .collect()
+        };
         match self {
             WithWitnesses(witnesses) if witnesses.is_empty() => WithWitnesses(witnesses),
             WithWitnesses(witnesses) => {
-                let new_witnesses = if matches!(ctor, Constructor::Missing) {
-                    let mut split_wildcard = SplitWildcard::new(pcx);
-                    split_wildcard.split(pcx, matrix.head_ctors(pcx.cx));
-                    // Construct for each missing constructor a "wild" version of this
-                    // constructor, that matches everything that can be built with
-                    // it. For example, if `ctor` is a `Constructor::Variant` for
-                    // `Option::Some`, we get the pattern `Some(_)`.
-                    let new_patterns: Vec<_> = split_wildcard
-                        .iter_missing(pcx)
-                        .map(|missing_ctor| {
-                            Fields::wildcards(pcx, missing_ctor).apply(pcx, missing_ctor)
-                        })
-                        .collect();
-                    witnesses
-                        .into_iter()
-                        .flat_map(|witness| {
-                            new_patterns.iter().map(move |pat| {
-                                let mut witness = witness.clone();
-                                witness.0.push(pat.clone());
-                                witness
+                let new_witnesses = match ctor {
+                    // This means we have a `non_exhaustive` enum missing a pattern
+                    //
+                    // This protects `match non_exhaustive_enum {}` where there are no patterns
+                    // and the error is "pattern `_` not covered", it is needed :(
```

Comment:
I was trying to comment on why the branches are needed, if we have a non_exhaustive enum we need to use the proper `Constructor::NonExhaustive` or some of the lint messages are wrong.

```rust
#[non_exhaustive]
pub enum Foo { A, B, C }

// another crate
match Foo::A {}
// fails and the `Witness` needs to have `Constructor::NonExhaustive` to create the correct message
```
I was at one point doing the same thing for `nonexhaustive_enum_missing_real_variants` being true or false and it failed other UI tests because of the wrong message

Reply:
I've made another suggestion